### PR TITLE
Change rule type to use polymorphic variants

### PR DESCRIPTION
**Pull request by kopecs**
Instead of using a normal variant for the mode field this changes the
rule type (`Rule.rule` -> `'mode Rule.rule_info`; `Rule.rule` is now an
alias for this with `'mode` as either a polymorphic variant for search
or taint) to be parameterised over a type variable `'mode` which is
instantiated to be specific polymorphic variants depending on the mode
of the rule.

This allows us to have functions which take/return rules where the modes
are restricted to a subset of all possible modes. Having this capability
is useful for if/when we have additional modes and, e.g., we want to
have a function take both search and taint rules but not rules having
other modes.

This just does the initial refactoring to change the main rule type and
some low-hanging fruit around the
`Match_{search,tainting}_mode.check_rule` types. There is probably some
additional refactoring which can be done to avoid redundant matches.

PR checklist:

- [x] Documentation is up-to-date
- [x] Changelog is up-to-date
- [x] Change has no security implications (otherwise, ping security team)

### CODE REVIEW

File: semgrep-core/src/core/Rule.ml
Comment (mmcqd) [inline semgrep-core/src/core/Rule.ml]: Having to write all these helpers and keep them updated is pretty annoying, wonder if there's a ppx deriving extension that will do this?

File: semgrep-core/src/core/Rule.ml
Comment (mmcqd) [inline semgrep-core/src/core/Rule.ml]: Don't need to fix it in this PR but this function seems kind of pointless now, since it basically just renames constructors

File: semgrep-core/src/engine/Match_rules.ml
Comment (mmcqd) [inline semgrep-core/src/engine/Match_rules.ml]: You don't need `as r` here because `r` is already bound to the rule

File: semgrep-core/src/metachecking/Check_rule.ml
Comment (mmcqd) [inline semgrep-core/src/metachecking/Check_rule.ml]: same deal with the `as r`

**Review by mmcqd — COMMENTED** [review]
Will wait on others for approval, but generally looks good to me. I think this doesn't make much of a difference right now, but I can imagine the usefulness once we add your Extract mode and we start needing non singleton subsets of rule modes as arguments.

File: semgrep-core/src/engine/Match_rules.ml
Comment (kopecs) [reply semgrep-core/src/engine/Match_rules.ml]: You do, because otherwise the compiler doesn't know we're in the specific variant.

File: semgrep-core/src/metachecking/Check_rule.ml
Comment (kopecs) [reply semgrep-core/src/metachecking/Check_rule.ml]: Ditto

File: semgrep-core/src/core/Rule.ml
Comment (kopecs) [reply semgrep-core/src/core/Rule.ml]: I think it's still useful, because it lets us go from a mixed list to two unmixed lists (N-many for N modes).

Comment (mmcqd) [reply semgrep-core/src/core/Rule.ml]: actually just kidding I did not know what `partition` does lol

File: semgrep-core/src/core/Rule.ml
Comment (kopecs) [reply semgrep-core/src/core/Rule.ml]: I don't think there would be, and if we have some fields only shared for a couple modes then we'll probably have to do something like this anyway unfortunately. I don't think it'll be too burdensome though.

Comment (IagoAbal) [reply semgrep-core/src/core/Rule.ml]: It may be worth considering having a `common : rule -> common` function, and then write all the functions below in terms of that, e.g. `let message rule = (common rule).message`.

Comment (kopecs) [reply semgrep-core/src/core/Rule.ml]: My only concern there would be wrt how "common" some of the data search and taint share would be to other modes (e.g., `severity`, `fix`)

**Comment by aryx** [comment]
meh, not convinced.

File: semgrep-core/src/core/Rule.ml
Comment (aryx) [reply semgrep-core/src/core/Rule.ml]: We could also make a polymorphic rule_
type 'mode rule_ = {
   id: ...
   message:
   mode: 'mode;
 }
and then define the mode = `Tainting of ... | `Search of ...
and then you can type rule = mode rule_
type tainting_rule = [`Tainting of ...] rule_
?

Comment (IagoAbal) [reply semgrep-core/src/core/Rule.ml]: I agree with @aryx here it's worth trying this. I believe subtyping also applies when using polymorphic variants as type parameters.

Comment (kopecs) [reply semgrep-core/src/core/Rule.ml]: > I believe subtyping also applies when using polymorphic variants as type parameters.

I couldn't find docs which mention variance explicitly, but a couple examples in the repl seemed to be fine, so I'm thinking the same.

Going to implement it like suggested above and see how much better/worse it is.

**Review by aryx — APPROVED** [review]

**Comment by aryx** [comment]
better

**Pull request merged**
